src/store/aggregator/aggregator.sagas.ts

```diff
@@ -169,7 +182,84 @@ export function* fetchPausedTokens() {
 
     yield* put(aggregatorActions.setIsStartingTokenPaused(pausedTokens));
   } catch (e) {
-    // console.error(e);
+    yield* put(aggregatorActions.fetchFeesAndLimitsFailure);
+  }
+}
+
+export function* fetchIncentive() {
+  try {
+    yield* put(aggregatorActions.setIncentivesLoading());
+
+    const sendAmount = yield* select(sendAmountSelector);
+    if (Number(sendAmount) === 0) return;
+    const flowState = yield* select(flowStateSelector);
+
+    const startingTokenAddress = yield* select(startingTokenAddressSelector);
+    const destinationTokenAddress = yield* select(
+      destinationTokenAddressSelector
+    );
+    if (!startingTokenAddress || !destinationTokenAddress) return;
+
+    let incentive = BigNumber.from(0);
+    let receiveAmount = BigNumber.from(0);
+    let incentiveType = IncentiveType.none;
+
+    if (flowState === 'deposit') {
+      const startingToken = yield* select(startingTokenSelector);
+      const bridge = yield* select(bridgeSelector);
+      const sovTokenAddress = bridge
+        ? bridge.getRskSovrynTokenAddress(startingToken!)
+        : destinationTokenAddress;
+
+      incentiveType = IncentiveType.reward;
+      const tokenDecimals = yield* select(startingTokenDecimalsSelector);
+      const amount = utils.parseUnits(sendAmount ?? '', tokenDecimals);
+      const destinationChain = yield* select(destinationChainSelector);
+      const isMainnet = destinationChain === ChainEnum.RSK;
+
+      incentive = (yield* call(
+        getReward,
+        sovTokenAddress!,
+        amount,
+        isMainnet
+      )) as BigNumber;
+      receiveAmount = amount.add(incentive);
+    } else if (flowState === 'withdraw') {
+      const destinationToken = yield* select(startingTokenSelector);
+      const bridge = yield* select(bridgeSelector);
+      const sovTokenAddress = bridge
+        ? bridge.getRskSovrynTokenAddress(destinationToken!)
+        : destinationTokenAddress;
+
+      incentiveType = IncentiveType.penalty;
+      const amount = utils.parseUnits(sendAmount ?? '', DEFAULT_ASSET_DECIMALS);
+      const startingChain = yield* select(startingChainSelector);
+      const isMainnet = startingChain === ChainEnum.RSK;
+
+      incentive = (yield* call(
+        getPenalty,
+        sovTokenAddress!,
+        amount,
+        isMainnet
+      )) as BigNumber;
+      receiveAmount = amount.sub(incentive);
+    }
+
+    //incentive = incentive && roundBN(incentive, 3);
+
+    yield* put(
+      aggregatorActions.setIncentives({
+        type: incentiveType,
+        amount: utils.formatUnits(incentive, DEFAULT_ASSET_DECIMALS),
+      })
+    );
+    yield* put(
+      aggregatorActions.setReceiveAmount(
```

Comment:
I think that this leads to bugs - you need to set receive amount also in case there is no reward/penalty because only `RSK -> RSK` conversions will work.

Reply:
I don't understand why you think so. Non rsk-rsk transactions also have reward/penalty.